File: backend/common/src/main/java/com/system/alba/model/ResponseModel.java
package com.system.alba.model;

import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.http.ResponseEntity;

@Getter
@NoArgsConstructor
public class ResponseModel<T> {

    private String code;
    private String message;
    private T data;

    public ResponseModel(T data) {
        this.data = data;
    }

    public static <T> ResponseModel<?> error(String code, String message) {
        ResponseModel<T> responseModel = new ResponseModel<>();
        responseModel.code = code;
        responseModel.message = message;
        return responseModel;
    }

    public static <T> ResponseEntity<ResponseModel<T>> ok(T data) {
        ResponseModel<T> responseModel = new ResponseModel<>();
        responseModel.data = data;
        return ResponseEntity.ok(responseModel);
    }
}

File: mobile-monorepo/packages/common/lib/src/model/response_model.dart
class ResponseModel<T> {
  const ResponseModel({
    this.code,
    this.message,
    this.data,
  });

  final String? code;
  final String? message;
  final T? data;

  factory ResponseModel.fromJson(
    Map<String, dynamic> json,
    T Function(Object? value)? fromJsonT,
  ) {
    return ResponseModel(
      code: json['code'] as String?,
      message: json['message'] as String?,
      data: json.containsKey('data') && fromJsonT != null
          ? fromJsonT(json['data'])
          : null,
    );
  }
}
